envoy/server.go

```diff
@@ -0,0 +1,437 @@
+	hcm "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	tcpp "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/tcp_proxy/v2"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	xds "github.com/envoyproxy/go-control-plane/pkg/server"
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/relistan/go-director"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+const (
+	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+	// TODO: Move this code in a common module and de-duplicate in sidecarhttp/envoy_api.go
+	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+	// Used to join service name and port. Must not occur in service names
+	ServiceNameSeparator = ":"
+)
+
+type xdsCallbacks struct{}
+
+func (*xdsCallbacks) OnStreamOpen(context.Context, int64, string) error  { return nil }
+func (*xdsCallbacks) OnStreamClosed(int64)                               {}
+func (*xdsCallbacks) OnStreamRequest(int64, *api.DiscoveryRequest) error { return nil }
+func (*xdsCallbacks) OnStreamResponse(_ int64, req *api.DiscoveryRequest, _ *api.DiscoveryResponse) {
+	if req.GetErrorDetail().GetCode() != 0 {
+		log.Errorf("Received Envoy error code %d. Check Envoy logs for details.", req.GetErrorDetail().GetCode())
+		log.Debugf("Envoy error: %s", req.GetErrorDetail().GetMessage())
+	}
+}
+func (*xdsCallbacks) OnFetchRequest(context.Context, *api.DiscoveryRequest) error   { return nil }
+func (*xdsCallbacks) OnFetchResponse(*api.DiscoveryRequest, *api.DiscoveryResponse) {}
+
+// Server is a wrapper around Envoy's control plane xDS gRPC server and it uses
+// the Aggregated Discovery Service (ADS) mechanism.
+type Server struct {
+	Listener      *Listener
+	config        config.EnvoyConfig
+	state         *catalog.ServicesState
+	snapshotCache cache.SnapshotCache
+	xdsServer     xds.Server
+}
+
+// Run sets up the Sidecar listener event loop and starts the Envoy gRPC server
+func (s *Server) Run(ctx context.Context, looper director.Looper) {
+	grpcServer := grpc.NewServer()
+	discovery.RegisterAggregatedDiscoveryServiceServer(grpcServer, s.xdsServer)
+
+	go looper.Loop(func() error {
+		// Discard the event since we need a snapshot of the entire state
+		<-s.Listener.eventsChan
+
+		// When a server is expired in catalog/services_state.go -> ExpireServer(),
+		// the listener will receive an event for each expired service. We want to
+		// flush the channel to prevent rapid-fire updates to Envoy.
+		// This was inspired from receiver/receiver.go -> ProcessUpdates().
+		//
+		// TODO: We might want to consider flushing it in an infinite loop until
```

Comment:
Can we just buffer more events? 20 was used by the URL listener, but we are internal to the process and could buffer/flush more events faster.

Reply:
Since we actually regenerate the whole Envoy state on each "update" we do need to try pretty hard to de-dupe here. Your idea about an infinite loop might be the right direction. But maybe we say "loop until empty or 2 seconds have passed"

Reply:
We'll need to think harder about this. There is no reason for the "loop until empty" to not finish and it will finish really fast. However, something like `state.ExpireServer()` will be slower, since it performs more operations. Only a `time.Sleep(x)` with a small `x` would allow [this loop](https://github.com/Nitro/sidecar/blob/9e99a8496e065f2099a58618ebabb912a07ae02c/catalog/services_state.go#L172-L177) to finish processing after we receive the initial event from the first iteration, but then we'd add this artificial `x` delay before sending any update to Envoy. But maybe that's fine if `x` is less than 100 ms?

Reply:
OK, this is no longer relevant, since I changed the implementation to use a simple TimedLooper that runs every second instead of hooking into the listener mechanism.